### PR TITLE
Move the implementation of the code generation SPI

### PR DESCRIPTION
**Pull request by thobe**
Moves the knowledge of which APIs in the Kernel API that the generated code should call to the main cypher module, so that the 2.3 version of the language is not tied to a specific version of the internals of Neo4j

### CODE REVIEW

**Comment by thobe** [comment]
There is one downside to this change. It moves the testing of the generated code to the cypher/cypher module, meaning that code that exists in cypher/compiler isn't tested in that module.

I don't know how to solve that though. - other than having some different implementation of `CodeStructure` that performs verification of the IR rather than generating code.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/LogicalPlanningTestSupport.scala [hunk 1]
```diff
@@ -165,7 +165,7 @@ trait LogicalPlanningTestSupport extends CypherTestSupport with AstConstructionT
       queryPlanner = queryPlanner,
       rewriterSequencer = rewriterSequencer,
       plannerName = None,
-      runtimeBuilder = SilentFallbackRuntimeBuilder(InterpretedPlanBuilder(Clock.SYSTEM_CLOCK, monitors), CompiledPlanBuilder(Clock.SYSTEM_CLOCK)),
+      runtimeBuilder = InterpretedRuntimeBuilder(InterpretedPlanBuilder(Clock.SYSTEM_CLOCK, monitors)),
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/LogicalPlanningTestSupport.scala:168]: I was somewhat concerned over this change, but then I checked and we never actually use `def newPlanner` so I guess an alternative to this change would be to just remove the method.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/GeneratedQueryStructure.scala [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+    returns(get(self(), typeRef[ExecutionMode], "executionMode")).
+    build()
+  val EXECUTION_PLAN_DESCRIPTION = MethodTemplate.method(typeRef[InternalPlanDescription], "executionPlanDescription").
+    returns(invoke(get(self(), typeRef[Supplier[InternalPlanDescription]], "description"), method[Supplier[InternalPlanDescription], InternalPlanDescription]("get"))).
+    build()
+  val JAVA_COLUMNS = MethodTemplate.method(typeRef[util.List[String]], "javaColumns").
+    returns(get(typeRef[util.List[String]], "COLUMNS")).
+    build()
+}
+
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/GeneratedQueryStructure.scala]: At some point we should break up this file, it is getting longer by the minute.  Not necessary right now though.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/LogicalPlanningTestSupport.scala (hunk 1 above)
Comment (thobe) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/LogicalPlanningTestSupport.scala:168]: so was I, but now that you tell me it isn't used, I'm not concerned anymore :)

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/GeneratedQueryStructure.scala (hunk 2 above)
Comment (thobe) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/GeneratedQueryStructure.scala]: I agree.

**Pull request merged**